MIG stock_mts_mto_rule to 12.0

jaredkipe opened the pull request:
There were major changes to the way Warehouses maintain their routes/rules and self heal in 12.0 

This necessitated large changes to `models/stock_warehouse.py` and some changes to test expectations.

In general, I read through the new Odoo code to get a feel for how they are updating and healing routes, then I applied the same methodology to the refactor work. I didn't really comment these methods as a search through **purchase, mrp, stock** for the same method names will turn up very similar looking code (that is also not commented).

Some not so great things...
In the model `stock.warehouse` Odoo provides a shim for an old method named `get_all_routes_for_wh` that is now effectively named `_get_all_routes`.  However, in **purchase_stock** module, they incorrectly call the parent method that isn't actually the method being called.

```
    @api.multi
    def _get_all_routes(self):
        routes = super(StockWarehouse, self).get_all_routes_for_wh()
```

This can lead to infinite loops, so ultimately I ended up just resorting to the same 'shim' approach that Odoo took.

florian-dacosta committed:
refactore creating a new route, a new procurement action and separate procurement rules to handle the mts+mto scenario

florian-dacosta committed:
add test + use standard mts and mto rules

florian-dacosta committed:
fix stock_mts_mto procurement check method

gabrielo77 committed:
Better readme in stock_mts_mto_rule

oca-transbot committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
[UPD] prefix versions with 8.0

pedrobaeza committed:
[MIG] Make modules uninstallable

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
Update READMEs

Conforms to the latest README template: bugtracker, runbot etc.
Fixes bugtracker URL on some modules.
States OCA as maintainer, removes other contributors from the 'Maintainer' section.

JulioSerna committed:
set pull rule of customer location like split_procurement to allow when warehouse is two/three steps delivery propagate move by mts-mto in pick/pack/out

oca-transbot committed:
OCA Transbot updated translations from Transifex

OCA Transbot updated translations from Transifex

lk-eska committed:
stock move name fix

otherjustin committed:
Update README.rst

Add explanation for finding this setting

florian-dacosta committed:
Do no run mts mto procurement if it already has been splited into at least another procurement

JordiBForgeFlow committed:
stock_mts_mto_rule

sbidoul committed:
[FIX] remove en.po that was erroneously created by transbot

pedrobaeza committed:
[MIG] Rename manifest files

Commit:
[MIG] Migrate stock_mts_mto_rule to v10.0

Commit:
[IMP] stock_mts_mto_rule : Added some unit tests

Commit:
[FIX] stock_mts_mto_rule : Fixed v10 migration bugs, revealed by new unit tests

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] stock_mts_mto_rule: Don't fail on renaming WH code

There's a traceback when renaming warehouse code (without renaming warehouse name), as name argument is False.

oca-transbot committed:
OCA Transbot updated translations from Transifex

OCA Transbot updated translations from Transifex

OCA Transbot updated translations from Transifex

zaoral committed:
[FIX] _update_routes now handle multiple records (#325)

This method is called from the write() method the one how support
multiple records an api.multi method.

I was making some test when I tried to activate the mto+mts option for
multiple warehouse and the next error appears: ``ValueError: Expected singleton``

In order to fix this error I only added a loop to manage the multiple
registers.

oca-transbot committed:
OCA Transbot updated translations from Transifex

OCA Transbot updated translations from Transifex

florian-dacosta committed:
Migrate stock_mts_mto_rule to version 11

oca-travis committed:
[UPD] Update stock_mts_mto_rule.pot

OCA-git-bot committed:
[ADD] icon.png

jaredkipe committed:
[MIG] Migrate stock_mts_mto_rule to 12.0

Major changes to the way Warehouses update their routes/rules and self heal.
Improved rule code to use `float_compare` and `float_is_zero`.

manney commented:
We've just pulled this branch.  I did a quick functional test on it as well as ran all of the tests.  Everything seems to pass with flying colors.  :+1:

jaredkipe commented:
@manney thank you!

Anyone else interested in trying it, and potentially other beta branches like `connector` is welcome to use our automated docker builds at `registry.gitlab.com/hibou-io/hibou-odoo/suite:12.0-test`

rousseldenis commented on stock_mts_mto_rule/__init__.py:
```diff
@@ -0,0 +1 @@
+from . import model
```
models
https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#directories

rousseldenis commented on stock_mts_mto_rule/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+{'name': 'Stock MTS+MTO Rule',
+ 'version': '12.0.1.0.0',
+ 'author': 'Hibou Corp.,Akretion,Odoo Community Association (OCA)',
+ 'website': 'http://www.akretion.com',
```
Could you follow OCA conventions for manifest (website, missing entires) ?

https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py

rousseldenis commented on stock_mts_mto_rule/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+ 'summary': 'Add a MTS+MTO route',
+ 'depends': ['stock',
+             ],
+ 'demo': [],
```
Please remove void one

rousseldenis commented on stock_mts_mto_rule/model/stock_rule.py:
```diff
@@ -0,0 +1,70 @@
+                if not rule.mts_rule_id or not rule.mto_rule_id:
+                    msg = _('No MTS or MTO rule configured on procurement '
+                            'rule: %s!') % (rule.name, )
+                    raise UserError(msg)
```
Constrains must raise ValidationError

rousseldenis commented on stock_mts_mto_rule/model/stock_rule.py:
```diff
@@ -0,0 +1,70 @@
+                    msg = _('Inconsistency between the source locations of '
+                            'the mts and mto rules linked to the procurement '
+                            'rule: %s! It should be the same.') % (rule.name,)
+                    raise UserError(msg)
```
Constrains must raise ValidationError

rousseldenis commented on stock_mts_mto_rule/model/stock_rule.py:
```diff
@@ -0,0 +1,70 @@
+        qty_available = product.uom_id._compute_quantity(
+            virtual_available, product_uom)
+
+        if qty_available > 0:
```
You should use float_compare

rousseldenis commented on stock_mts_mto_rule/model/stock_rule.py:
```diff
@@ -0,0 +1,70 @@
+        needed_qty = self.get_mto_qty_to_order(product_id, product_qty,
+                                               product_uom, values)
+
+        if needed_qty == 0.0:
```
You should use float_is_zero

rousseldenis commented on stock_mts_mto_rule/model/stock_rule.py:
```diff
@@ -0,0 +1,70 @@
+            getattr(self.mts_rule_id, '_run_%s' % self.mts_rule_id.action)(
+                product_id, product_qty, product_uom, location_id, name,
+                origin, values)
+        elif needed_qty == product_qty:
```
You should use float_compare

rousseldenis commented on stock_mts_mto_rule/model/stock_warehouse.py:
```diff
@@ -0,0 +1,95 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields
```
```suggestion
from odoo import fields, models, _
```

rousseldenis reviewed (changes requested):
Code review.

florian-dacosta commented on stock_mts_mto_rule/README.rst:
```diff
@@ -0,0 +1,92 @@
+.. image:: https://img.shields.io/badge/license-AGPL--3-blue.png
```
It would be nice to split the readme, like explain here : https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-12.0

florian-dacosta commented on stock_mts_mto_rule/README.rst:
```diff
@@ -0,0 +1,92 @@
+------------
+
+* Florian da Costa <florian.dacosta@akretion.com>
+
```
You can add yourself in the contributor list, but not in the authors

florian-dacosta commented on stock_mts_mto_rule/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+
+{'name': 'Stock MTS+MTO Rule',
+ 'version': '12.0.1.0.0',
+ 'author': 'Hibou Corp.,Akretion,Odoo Community Association (OCA)',
```
You should put yourself in the contributor and not in the authors

florian-dacosta reviewed (changes requested):
Thanks for the migration!
I made some minor comments
Otherwise I did functional tests and it seems to work as expected.

rousseldenis commented:
@jaredkipe Do you plan to attend comments?

Jacekdaa commented:
nothing specific. at this moment he will give a git chada and honestly say little wien.

rousseldenis commented:
> nothing specific. at this moment he will give a git chada and honestly say little wien.

Ugh, didn't understand...

jaredkipe commented:
Now that Odoo has merged in https://github.com/odoo/odoo/pull/33288 , I have removed the patch and re-worked the test.

Otherwise I believe I have addressed everyone's concerns other than splitting the README into multiple files.

rousseldenis commented:
@jaredkipe Just lint errors and LGTM!

jaredkipe commented:
Should be good now, feel free to squash as there are now a couple of simple 'fix' commits that don't really help.

rousseldenis commented:
> Should be good now, feel free to squash as there are now a couple of simple 'fix' commits that don't really help.

Hi Jared, I cannot just squash your four migration commits during merging. Could you do it ?

jaredkipe commented:
All done with that, the only major difference is that it now thinks the `stock_warehouse.py` file is a complete rewrite due to the changes and `model` => `models` module rename.

sya-odoo commented:
Hi Folks, any idea when this PR is likely to be approved? we do have a customer waiting on this - blocking issue for him & his business. Would be very much appreciated, thanks a lot in advance! :-)

rousseldenis reviewed (approved):
Code review. LGTM

rousseldenis commented:
/ocabot merge
